### ISSUE TITLE
"bvh2features.py" what exactly returns?

### ISSUE THREAD

**Issue by Famosi** [Svito-zar/gesticulator#11]
Hi @Svito-zar, I'm working on a speech-to-gesture model for social robots. The goal of the project is to learn a model that predicts gestures from a given audio. Then, through a "motion2robot" mapping, the robot imitates predicted gestures.
 
Currently what I need to do is to extract features from `BVH` files. I've downloaded the _GENEA2020_ and _trinityspeechgesture_ datasets. At this step, I think your `bvh2features.py` script could come to my aid. Here I have some questions on this script:

1. The [description of the script](https://github.com/Svito-zar/gesticulator/blob/11642221751593e801790907b0dd5247d88b6468/gesticulator/data_processing/bvh2features.py#L5) says "_This script converts a gesticulation dataset from the BVH format to joint angles._". I can't get what exactly "joint angles" are. For example, If I have for the _joint_i_ `[Zrotation_i Xrotation_i Yrotation_i]` rotation coordinates, what I have as result from the script for the _joint_i_? How can I interpret these new angles? 

2. I need only the upper body. Is this what [`JointSelector`](https://github.com/Svito-zar/gesticulator/blob/11642221751593e801790907b0dd5247d88b6468/gesticulator/data_processing/bvh2features.py#L43) allows me to do? 

3. You selected 9 joints (`['Head','RightShoulder', 'RightArm', 'RightForeArm', 'RightHand', 'LeftShoulder', 'LeftArm', 'LeftForeArm', 'LeftHand']`). I've tried to give in input to this script a BVH file with `fps=60` and I have in output a numpy array with shape `(n_frame, 27)`. From this shape I deduce that I have, for each Joint, 3 values. What this values are is the answer to 1. but, if the deduction is correct, what's the order of this array? Is the order the same as the declaration in the `JointSelector`? For example, in this case, does `[npy_array[0][0], npy_array[0][1], npy_array[0][2]` correspond to the `Head` angles at `frame=0`?

**P.S**: My model will use only audio features (no text). Is there a way to train your model using only audio?

Many thanks for your help and congrats for your works :)

**Comment by FaMoSi** [comment]
I updated the files to reflect #462. The PR is #472.
Let me know if everything is ok ✅. 
I'm participating to the `hacktoberfest2020`. Please, add the label `hacktoberfest-accepted` 😀

**Comment by Svito-zar** [comment]
Hi @FaMoSi . Thank you for your questions.

Let me go through them one by one:

1. The description should be actually updated. The `bvh2features.py` script converts motion into expontial maps format. You can read more about expontial maps representation is this Wikipedia article: https://en.wikipedia.org/wiki/Axis–angle_representation 
2. Yes, that's right, joint selector let's you choose only the upper body. Actually we used only the upper body in our experiments anyway.
3. For this question, I am not sure what the answer is. Maybe my co-author, @simonalexanderson could answe this.

As for the last comment - you can remove text information from the model if you want to, but as you can see in our paper and videos our model would not perform well without text. If you want to use only audio - I would recommend you to rather use some of our previous models:
https://github.com/GestureGeneration/Speech_driven_gesture_generation_with_autoencoder
https://github.com/simonalexanderson/StyleGestures

**Comment by Famosi** [comment]
Thanks @Svito-zar for your precise and quick response!  

I'll really appreciate if @simonalexanderson could help me to understand how to interpret the output of the `bvh2features.py` script. In particular, my goal is to transform this _features_ into robot's joint angles (Yaw, Pitch, Roll).

Regarding the "only-audio" model, I'll check and try out your previous models.

**Comment by simonalexanderson** [comment]
Hi @FaMoSi,

The best way to check the order of the features is to remove the last preprocessing step, called Numpyfier, in the data-pipeline. The features will then be expressed in pandas format, and you can check the feature names by printing data.values.columns. This will look something like this:

```
 data_pipe = Pipeline([
       ('dwnsampl', DownSampler(tgt_fps=fps,  keep_all=False)),
       ('root', RootTransformer('hip_centric')),
       ('mir', Mirror(axis='X', append=True)),
       ('jtsel', JointSelector(['Spine','Spine1','Spine2','Spine3','Neck','Neck1','Head','RightShoulder', 'RightArm', 'RightForeArm', 'RightHand', 'LeftShoulder', 'LeftArm', 'LeftForeArm', 'LeftHand'], include_root=True)),
       ('exp', MocapParameterizer('expmap')), 
       ('cnst', ConstantsRemover())
    ])

    out_data = data_pipe.fit_transform(data_all)
    print (out_data[0].values.columns)
```

I beleive they will be in the same order as in the JointSelector, but check just in case.

**Comment by Famosi** [comment]
_-- In case it will be useful for someone else --_

The order is not the same as in the `JointSelector`. Here is an example:

**JointSlector**: 
```
['Spine','Spine1','Spine2','Spine3','Neck','Neck1','Head','RightShoulder', 'RightArm', 'RightForeArm', 'RightHand', 'LeftShoulder', 'LeftArm', 'LeftForeArm', 'LeftHand']
```

**Output** of `out_data[0].values.columns`: 
```
['LeftHand_alpha', 'LeftHand_beta', 'LeftHand_gamma',
       'LeftForeArm_alpha', 'LeftForeArm_beta', 'LeftForeArm_gamma',
       'LeftArm_alpha', 'LeftArm_beta', 'LeftArm_gamma', 'LeftShoulder_alpha',
       'LeftShoulder_beta', 'LeftShoulder_gamma', 'RightHand_alpha',
       'RightHand_beta', 'RightHand_gamma', 'RightForeArm_alpha',
       'RightForeArm_beta', 'RightForeArm_gamma', 'RightArm_alpha',
       'RightArm_beta', 'RightArm_gamma', 'RightShoulder_alpha',
       'RightShoulder_beta', 'RightShoulder_gamma', 'Head_alpha', 'Head_beta',
       'Head_gamma', 'Neck1_alpha', 'Neck1_beta', 'Neck1_gamma', 'Neck_alpha',
       'Neck_beta', 'Neck_gamma', 'Spine3_alpha', 'Spine3_beta',
       'Spine3_gamma', 'Spine2_alpha', 'Spine2_beta', 'Spine2_gamma',
       'Spine1_alpha', 'Spine1_beta', 'Spine1_gamma', 'Spine_alpha',
       'Spine_beta', 'Spine_gamma']
```

Thanks for your help :)
